### PR TITLE
Re-enable Fused BatchNorm + Add + Activation for the backprop

### PR DESCRIPTION
**Pull request by kaixih**
This PR re-enables the fused (BatchNorm + Add + Activation) for the backprop. This is essentially a combination of the two previous PRs + one line change to disable the remapping when XLA is on:
1. Original PR: https://github.com/tensorflow/tensorflow/pull/48063 
2. Fix an issue of dual BN inputs: https://github.com/tensorflow/tensorflow/pull/48893
3. Disable the remapping when XLA is on.

cc. @nluehr

### CODE REVIEW

File: tensorflow/core/grappler/optimizers/remapper.cc
Comment (cheshire) [inline tensorflow/core/grappler/optimizers/remapper.cc:1028]: This is not enough though: there are other ways to enable XLA. Additionally, we would be to dig to the bottom of the issue and figure out why the convolution configuration was marked as unsupported and whether it's specific to XLA.

**Review by cheshire — CHANGES_REQUESTED** [review]

**Comment by cheshire** [comment]
Out of curiosity, what is the motivation for this PR? Grappler is not very well-maintained now. Is it possible to use XLA via autoclustering or jit_compile and do the rewrite there instead?

File: tensorflow/core/grappler/optimizers/remapper.cc
Comment (kaixih) [reply tensorflow/core/grappler/optimizers/remapper.cc:1028]: I agree. The reason I put a line here is that I heard the XLA might not be able to correctly process the new fused op. So, I tried to avoid remapping this pattern when XLA is on. For the "unsupported cudnn conv" issue, if you can share with more details on how to repro the failed model benchmark, I can look into it.

**Comment by cheshire** [comment]
I've replied to your other email. The issue has nothing to do with XLA, raw
TF code is making the failing call.

**Review by cheshire — CHANGES_REQUESTED** [review]
We need to figure out whether the regression is still there before we can land it. I don't think that Volkswagen-like tests will get us there.

**Comment by kaixih** [comment]
> Volkswagen-like tests

**Comment by kaixih** [comment]
Sorry clicked the wrong button.

**Comment by kaixih** [comment]
> We need to figure out whether the regression is still there before we can land it. I don't think that Volkswagen-like tests will get us there.

Hi @cheshire I am still working on the repro. This above commit is not to trigger the landing process. It is just to fix one bug we found internally when testing our models.

**Comment by kaixih** [comment]
Here is more background to show this is _not_ a "Volkswagen-like tests":

We hit this issue yesterday when training one of our JoC models: `InvalidArgumentError: FusedBatchNormGrad with activation is only supported when is_training=True` After some digging, I found that we accidentally use the fused batchnorm grad in the inference mode. However, the new remapping is supposed to be only working with training mode. Then, I realized that the remapper's FindXXX function should check the mode and if it is inference mode, we need to skip the remapping. After the fix, the issue is gone.

So, this has nothing to do with the repro you provided in the email and it has nothing to do to push the merge with this single commit. I am still working on that since it is not simple to get the benchmark compiled and run. I will let you know when we have progress there.

@sanjoy @cheshire @nluehr @cliffwoolley

File: tensorflow/core/ops/nn_ops.cc
Comment (cheshire) [inline tensorflow/core/ops/nn_ops.cc:326]: I understand is_training is present in the ops above already, but I'm still a bit confused what does this attribute mean for the op definition?

File: tensorflow/core/grappler/optimizers/remapper.cc
Comment (cheshire) [inline tensorflow/core/grappler/optimizers/remapper.cc:121]: I don't think this constructor definition does anything.

File: tensorflow/core/grappler/optimizers/remapper.cc
Comment (cheshire) [reply tensorflow/core/grappler/optimizers/remapper.cc:1050]: Right, but still action item here is to have a more robust check. In Python, you could check whether you are inside XLA context, I'm not 100% sure what is the best analogue in Grappler.

File: tensorflow/core/grappler/optimizers/remapper.cc
Comment (cheshire) [inline tensorflow/core/grappler/optimizers/remapper.cc:1056]: Code style is using less `auto` statements

**Review by cheshire — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/optimizers/remapper.cc
Comment (kaixih) [reply tensorflow/core/grappler/optimizers/remapper.cc]: Done.

File: tensorflow/core/grappler/optimizers/remapper.cc
Comment (kaixih) [reply tensorflow/core/grappler/optimizers/remapper.cc]: Removed.

File: tensorflow/core/ops/nn_ops.cc
Comment (kaixih) [reply tensorflow/core/ops/nn_ops.cc:326]: Yes, the attribute is from the above op `FusedBatchNormGradV3`. Both the `FusedBatchNormGradV3` and this new `_FusedBatchNormGradEx` shared the same base class `FusedBatchNormGradOpBase` which needs this attribute to determine to call normal `functor::FusedBatchNormGrad` or `functor::FusedBatchNormFreezeGrad`. And we don't expect `is_training=False` for `_FusedBatchNormGradEx` and only check it.

File: tensorflow/stream_executor/cuda/cuda_dnn.cc
Comment (kaixih) [reply tensorflow/stream_executor/cuda/cuda_dnn.cc]: Done.

File: tensorflow/stream_executor/cuda/cuda_dnn.cc
Comment (kaixih) [reply tensorflow/stream_executor/cuda/cuda_dnn.cc]: Done.

File: tensorflow/stream_executor/cuda/cuda_dnn.cc
Comment (kaixih) [reply tensorflow/stream_executor/cuda/cuda_dnn.cc]: Done.

File: tensorflow/core/kernels/fused_batch_norm_op.cc
Comment (kaixih) [reply tensorflow/core/kernels/fused_batch_norm_op.cc:1653]: I think originally we separate the PR to two commits for (1) Add _FusedBatchNormGradEx op (This is with the test on this new op.) and (2) Add remapper support. I merged them together in one commit in this PR since I think they have been reviewed and approved.
For the XLA lowering, our XLA team plans to add this op to their allowlist. Probably another PR. For this PR, we simply don't remap to this new op when XLA is detected to avoid XLA seeing it.

**Review by cheshire — CHANGES_REQUESTED** [review]
I don't think your check for XLA would work under jit_compile: could you test?

**Comment by kaixih** [comment]
> I don't think your check for XLA would work under jit_compile: could you test?

Sure. Do you mean the `tf.function(jit_compile=True)`? I've tried this code below, where I can see that if `tf.function` is applied, both the forward and backward passes will call the fused cudnn APIs. Instead, if the `@tf.function(jit_compile=True)` is used, there is no cudnn used for batch norm. I checked with our XLA team and learned that "there are no graph optimization passes run with jit_compile=True at TF level, only XLA optimizations used post conversion/during compilation". So, it sounds the jit_compile should be fine for this case.

```python
import tensorflow as tf
from tensorflow.keras import layers

N, H, W, C = 2, 2, 3, 3
tf.random.set_seed(1)


dense = layers.Dense(4)
conv2d = layers.Conv2D(4, 2, padding='same')
batch_norm = layers.BatchNormalization()
relu = layers.ReLU()

tf.keras.mixed_precision.set_global_policy('mixed_float16')

model = tf.keras.Sequential()
model.add(layers.Dense(4))
model.add(layers.Conv2D(4, 2, padding='same'))
model.add(layers.BatchNormalization())
model.add(layers.ReLU())

#@tf.function(jit_compile=True)
@tf.function
def train_step(x):
  with tf.GradientTape() as tape:
    y = model(x, training=True)
    loss = tf.reduce_mean(y)

  grads = tape.gradient(loss, model.trainable_variables)
  print(grads)

  return loss, grads

x = tf.random.uniform((N, H, W, C))
loss, grads = train_step(x)

print("Loss:", loss)
print("Grads:", grads)
```

**Comment by cheshire** [comment]
hi,

Unfortunately this is a known Grappler/XLA wart, what you need to do is to
wrap tf.function(jit_compile=True) in another (uncompiled) function, and
then you should see those grappler optimizations applied.

**Comment by kaixih** [comment]
That is good to know. But could you be more specific on what I should put to the jit_compile function and what to the uncompiled function? I tried the code below but it seems not right:
```python
@tf.function(jit_compile=True)
def train_step(x):
  with tf.GradientTape() as tape:
    y = model(x, training=True)
    loss = tf.reduce_mean(y)

  grads = tape.gradient(loss, model.trainable_variables)
  print(grads)

  return loss, grads

def train_step_wrapper(x):
  return train_step(x)
```

**Comment by cheshire** [comment]
Sorry, I've meant uncompiled *tf.function*.

If you add ***@***.***` on top of `train_step_wrapper`, you should see
grappler passes being applied.

Sorry for this design wart.

**Comment by kaixih** [comment]
Thanks. I've tried different combination of decorators (`tf.function(jit_compile)` and `tf.function`) over `train_step` and `train_step_wrapper` and all works fine. Specifically:

- jit_compile + jit_compile: no grappler (I think this is equivalent to a single function + jit_compile)
- jit_compile + tf.function: grappler is seen (I think since BN nodes has already been remapped during XLA phase, grappler's remapper doesn't take effects)
- tf.function + tf.function: grappler is seen (I think this is equivalent to a single function + tf.function)
- tf.function + jit_compile: no grappler (I think this is also equivalent to a single function + jit_compile)

Furthermore, I am curious what happens if only partial step (forward pass) is in a separate function, like:

```python
@tf.function
#@tf.function(jit_compile=True)
def forward(x):
  return model(x, training=True)

#@tf.function
@tf.function(jit_compile=True)
def train_step(x):
  with tf.GradientTape() as tape:
    y = forward(x)
    loss = tf.reduce_mean(y)

  grads = tape.gradient(loss, model.trainable_variables)
  print(grads)

  return loss, grads

x = tf.random.uniform((N, H, W, C))
loss, grads = train_step(x)
```
The results are 
- jit_compile + jit_compile: no grappler
- jit_compile + tf.function: grappler is seen (In this case, since forward BN node has already been remapped during XLA phase, grappler's remapper cannot see an eligible BN and doesn't take effects)
- tf.function + tf.function: grappler is seen
- tf.function + jit_compile: no grappler

Any advise?

**Comment by cheshire** [comment]
yeah unfortunately right now we have this implementation quirk, and we need
to make sure the code works with and without grappler (again, normally it's
fine, since for compiled code grappler does not add to performance).

The reason is that we use the "custom kernel" mechanism to execute the
jit_compile'd function, so if there's no outer function, no graph is
transformed (since we launch the compiled kernel straight from the eager
mode).

When there's an outer uncompiled function, we rewrite the graph first, and
only then launch the compiled subgraph.

In case of your PR it needs to work when the grappler is on and the scope
is compiled.

**Comment by kaixih** [comment]
I further modified the code as below which still works fine. Not sure if the `forward2` belongs to the "outer uncompiled function" you just mentioned. My intent is to make both jit and grapper remapper come into effects. So, the `forward` function is to use the "custom kernel" for BN and `forward2` will use fused cudnn BN after grappler. From the log, I can see that grapper remapper is on and successfully does the fusion.

```python
@tf.function(jit_compile=True)
def forward(x):
  return model(x, training=True)

def forward2(x):
  return model(x, training=True)

@tf.function
def train_step(x):
  with tf.GradientTape() as tape:
    y = forward(x)
    y = forward2(y)
    loss = tf.reduce_mean(y)

  grads = tape.gradient(loss, model.trainable_variables)
  print(grads)

  return loss, grads
```

**Comment by cheshire** [comment]
OK it's great that it works, but I am a bit confused about how though -
from what I understand, grappler produces the kernel XLA can not compile,
so what does it do?
Do you see the grappler rewrite kicking in from the logs?

**Comment by kaixih** [comment]
Yes, I can see the grappler rewrite logs for the above example:
```
2021-06-17 18:10:25.555918: I tensorflow/core/grappler/optimizers/remapper.cc:1699] Fuse FusedBatchNormGrad with ReluGrad:  fused_batch_norm_grad=gradient_tape/sequential/batch_normalization_1/FusedBatchNormGradV3 side_input=<none> activation=gradient_tape/sequential/re_lu_1/ReluGrad corresponding FusedBatchNorm=sequential/batch_normalization_1/FusedBatchNormV3
...
2021-06-17 18:10:25.556003: I tensorflow/core/grappler/optimizers/remapper.cc:1627] Fuse Relu with FusedBatchNorm: activation=sequential/re_lu_1/Relu side_input=<none> invalidated=<none> fused_batch_norm=sequential/batch_normalization_1/FusedBatchNormV3
```

I think I understand your concern. Let me reiterate what we wanted to do:
Let's assume we have a fuseable network:

- BN -> loss -> BN_Grad

So, applying the standalone jit or tf.function would generate the following respectively:

- **jit:** customBN -> loss -> customBN_Grad (OK)

- **grappler:** FusedBN -> loss -> FusedBN_Grad (OK)

However, when mixing the jit and tf.function, it might cause problem:
- **tf.function calls jit function**
  - **jit:** customBN -> loss -> customBN_Grad
  - **grappler:** customBN -> loss -> customBN_Grad (OK)

This scenario is probably fine because grappler doesn't know customBN and the customBN won't be remapped again.

- **jit function calls tf.function**
  - **grappler:** FusedBN -> loss -> FusedBN_Grad
  - **jit:** customBN -> loss -> FusedBN_Grad (No, this is WRONG)

This would be problematic, since FusedBN_Grad has been added to the XLA's allowlist and thus becomes a standalone op without its corresponding FusedBN. However, in real use case, we usually call the tf.function from a jit function, where the grappler pass will be disabled. And we will see the following:

- **jit function calls tf.function**
  - **grappler (disabled):** BN -> loss -> BN_Grad
  - **jit:** customBN -> loss -> customBN_Grad (OK)

So, I think the real question is how to repro the above scenario that can call a grappler-optimized tf.function in the jit function. Please let me know if my understanding is correct. And if yes, do you have any magic ways to do that?

**Comment by cheshire** [comment]
Hi Kaixi, sorry, I'm not sure I follow. Grappler should run first, only then followed by XLA compilation.

From what I understand, this PR generated `FusedBN` and `FusedBN_Grad` which do not have an XLA lowering. Is that correct? Or am I missing something?
How would these ops behave when the compilation is requested?

**Comment by kaixih** [comment]
> Hi Kaixi, sorry, I'm not sure I follow. Grappler should run first, only then followed by XLA compilation.
> 
> From what I understand, this PR generated `FusedBN` and `FusedBN_Grad` which do not have an XLA lowering. Is that correct? Or am I missing something?
> How would these ops behave when the compilation is requested?

Yes, I agree. If that happens, I think it would expose some problem (as I mentioned in the above comment that we don't want a standalone `FusedBN_Grad` to be there).

But the question is how to trigger this. Here is some conflicting comments I've got:
- You mentioned: "Grappler should run first, only then followed by XLA compilation."
- But I heard from our XLA team: "There are no graph optimization passes run with jit_compile=True at TF level, only XLA optimizations used post conversion/during compilation"

And all of the above experiments (trying different combinations and wrapper functions) are actually exploring how to achieve "Grappler should run first, only then followed by XLA compilation." The closest one in my opinion was the below one, because I thought the `forward` function would be applied with grappler and then be input to the jit scope. However, it turned out no grappler passes were applied. 
```python 
@tf.function
def forward(x):
  return model(x, training=True)

@tf.function(jit_compile=True)
def train_step(x):
  with tf.GradientTape() as tape:
    y = forward(x)
    loss = tf.reduce_mean(y)

  grads = tape.gradient(loss, model.trainable_variables)
  print(grads)

  return loss, grads
```

At this moment, I have a feeling that "Grappler should run first, only then followed by XLA compilation."  may not be easily achieved from python level and we might need to have a saved grappler optimized graph (that already contains both `FusedBN` and `FusedBN_Grad`) and then we apply XLA compilation (?).

**Comment by cheshire** [comment]
On Thu, Jun 17, 2021 at 11:54 AM Kaixi Hou ***@***.***> wrote:

> Hi Kaixi, sorry, I'm not sure I follow. Grappler should run first, only
> then followed by XLA compilation.
>
> From what I understand, this PR generated FusedBN and FusedBN_Grad which
> do not have an XLA lowering. Is that correct? Or am I missing something?
> How would these ops behave when the compilation is requested?
>
> Yes, I agree. If that happens, I think it would expose some problem (as I
> mentioned in the above comment that we don't want a standalone
> FusedBN_Grad to be there).
>
> But the question is how to trigger this. Here is some conflicting comments
> I've got:
>
>    - You mentioned: "Grappler should run first, only then followed by XLA
>    compilation."
>    - But I heard from our XLA team: "There are no graph optimization
>    passes run with jit_compile=True at TF level, only XLA optimizations used
>    post conversion/during compilation"
>
> > But I heard from our XLA team: "There are no graph optimization passes
run with jit_compile=True at TF level, only XLA optimizations used post
conversion/during compilation"

Could you point out where you've heard that? I think this only covers the
case where the compiled function is not nested.

>
>
> And all of the above experiments (trying different combinations and
> wrapper functions) are actually exploring how to achieve "Grappler should
> run first, only then followed by XLA compilation." The closest one in my
> opinion was the below one, because I thought the forward function would
> be applied with grappler and then be input to the jit scope. However, it
> turned out no grappler passes were applied
>
@tf.functiondef forward(x):
>   return model(x, training=True)
> @tf.function(jit_compile=True)def train_step(x):
>   with tf.GradientTape() as tape:
>     y = forward(x)
>     loss = tf.reduce_mean(y)
>
>   grads = tape.gradient(loss, model.trainable_variables)
>   print(grads)
>
>   return loss, grads.
>
> In this case the uncompiled function is within the compiled one, so no
passes would be applied.

I think I was confused by your previous email --- I thought you've meant
that the grappler did run?
Anyway, what happens when you do

```
@tf.function
def outer():
   train_step() // Up to you to pass or hardcode the parameter

@tf.function(jit_compile=True)
def train_step(..):
   ...
```

In my understanding, and from what I believe I saw from previous
experiments, it should trigger grappler. Please let me know if it doesn't.

> At this moment, I have a feeling that "Grappler should run first, only
> then followed by XLA compilation." may not be easily achieved from python
> level and we might need to have a saved grappler optimized graph (that
> already contains both FusedBN and FusedBN_Grad) and then we apply XLA
> compilation (?).
>
Yes, loading SavedModel with a compiled section should also trigger
grappler rewrites first.

**Comment by kaixih** [comment]
Oh, sorry about the confusion. Let's first focus on your pattern (which should be equivalent to case 2 in https://github.com/tensorflow/tensorflow/pull/49277#issuecomment-862615375):
```
@tf.function
def outer():
   train_step() // Up to you to pass or hardcode the parameter

@tf.function(jit_compile=True)
def train_step(..):
   ...
```
Yes, still the grappler is detected in the log with TF_CPP_VMODULE=meta_optimizer=3. I also print out the nodes that the grappler can see:
```
ZZZ node: _Retval
ZZZ node: _Retval
ZZZ node: _Retval
ZZZ node: _Retval
ZZZ node: _Retval
ZZZ node: _Retval
ZZZ node: _Retval
ZZZ node: StatefulPartitionedCall
ZZZ node: _Arg
ZZZ node: _Arg
ZZZ node: _Arg
ZZZ node: _Arg
ZZZ node: _Arg
ZZZ node: _Arg
ZZZ node: _Arg
ZZZ node: _Arg
ZZZ node: _Arg

```
Originally, I put a comment:
> I think since BN nodes has already been remapped during XLA phase, grappler's remapper doesn't take effects

But, I feel my original comment for this might be wrong? The reason why grappler's remapper doesn't take effects is not because XLA has already been applied, but because the wrapped function becomes a StatefulPartitionedCall?

**Comment by cheshire** [comment]
yeah but then grappler is supposed to go into the call anyway, but i
suspect it does not, since the call is not inlined. I did not realize that.

OK then, it seems the only time your rewrite can interfere with XLA is via
autoclustering, and you do check for that.
Let's approve and see if anything breaks.

**Review by cheshire — APPROVED** [review]

**Review by Neilblaze — APPROVED** [review]

**Comment by kaixih** [comment]
Rebased to resolve a conflict.

**Review by pkanwar23 — APPROVED** [review]

**Comment by cheshire** [comment]
This now has the following failure: Failure: F0709 00:43:15.026466    2393 shape_inference.cc:114] Check failed: output(i).IsSet() 5 for {{node fused_batch_norm_grad}} = _FusedBatchNormGradEx[T=DT_HALF, U=DT_FLOAT, activation_mode="Relu", data_format="NHWC", epsilon=0.1, is_training=true, num_side_inputs=1, _device="/job:localhost/replica:0/task:0/device:GPU:0"](output_grad_cast, input_cast, scale, fused_batch_norm:3, fused_batch_norm:4, fused_batch_norm:5, offset, fused_batch_norm)

**Comment by kaixih** [comment]
> This now has the following failure: Failure: F0709 00:43:15.026466 2393 shape_inference.cc:114] Check failed: output(i).IsSet() 5 for {{node fused_batch_norm_grad}} = _FusedBatchNormGradEx[T=DT_HALF, U=DT_FLOAT, activation_mode="Relu", data_format="NHWC", epsilon=0.1, is_training=true, num_side_inputs=1, _device="/job:localhost/replica:0/task:0/device:GPU:0"](output_grad_cast, input_cast, scale, fused_batch_norm:3, fused_batch_norm:4, fused_batch_norm:5, offset, fused_batch_norm)

Emm... This should be straightforward to change. Originally, I shared the `FusedBatchNormGradShape` function for the new fused op, which doesn't set the output 5 - side_input_backprop. Let me work on a fix.

**Pull request merged**
